Added Tests for Query with inner query

oskardudycz opened the pull request:
Added tests for issue https://github.com/JasperFx/marten/issues/1105

mysticmind commented on src/Marten.Testing/Linq/query_with_inner_query.cs:
```diff
@@ -0,0 +1,119 @@
+            var serializer = new JsonNetSerializer();
+            serializer.Customize(s =>
+            {
+                s.TypeNameHandling = Newtonsoft.Json.TypeNameHandling.None;
```
I believe the same can be applied as an attribute on the IEnumerable properties on the class to achieve the same rather than customizing the serializer at the store level. Thoughts?

oskardudycz replied:
@mysticmind good catch, I'll add next case checking that

robert-skarzycki commented:
@oskardudycz, what about one more level down, i.e. having creation of test data looking like this:
```
 public static TypeWithInnerCollections Create(params string[] array)
        {
            return new TypeWithInnerCollections()
            {
                Id = Guid.NewGuid(),
                Array = array,
                Flatten = array.Aggregate((i, j) => i + j),
                Enumerable = array.AsEnumerable(),
                IEnumerableFromArray = array,
                IEnumerbaleFromList = array.ToList(),
                List = array.ToList(),
                IList = array.ToList(),
                IReadonlyCollection = array.ToList(),
                IReadonlyCollectionOfInnerClasses = new List<TypeWithInnerCollections>
                    {
                        new TypeWithInnerCollections()
                        {
                            Id = Guid.NewGuid(),
                            Array = array,
                            Flatten = array.Aggregate((i, j) => i + j),
                            Enumerable = array.AsEnumerable(),
                            IEnumerableFromArray = array,
                            IEnumerbaleFromList = array.ToList(),
                            List = array.ToList(),
                            IList = array.ToList(),
                            IReadonlyCollection = array.ToList(),
                            IReadonlyCollectionOfInnerClasses = new List<TypeWithInnerCollections>
                            {
                                new TypeWithInnerCollections()
                                {
                                    Id = Guid.NewGuid(),
                                    Array = array,
                                    Flatten = array.Aggregate((i, j) => i + j),
                                    Enumerable = array.AsEnumerable(),
                                    IEnumerableFromArray = array,
                                    IEnumerbaleFromList = array.ToList(),
                                    List = array.ToList(),
                                    IList = array.ToList(),
                                    IReadonlyCollection = array.ToList(),
                                }
                            }
                        }
                    }
            };
```
?

Then having a query over this extra-level leaves would fail:
```
await query.Query<TypeWithInnerCollections>()
                .Where(x => x
                .IReadonlyCollectionOfInnerClasses
                .Any(y => y.IReadonlyCollectionOfInnerClasses
                .Where(e => e.Flatten == "onetwo")
                .Any())
                || x
                .IReadonlyCollectionOfInnerClasses
                .Any(y => y.IReadonlyCollectionOfInnerClasses
                .Where(e => e.Flatten == "twothree").Any())
                )
                .ToListAsync();
```
with exception stating that only `Contains` is allowed inside `Any`.

P.S. That's what I meant by asking about subqueries on Gitter: https://gitter.im/JasperFx/marten?at=5c66d8b64bafd97ca3f8f457

oskardudycz commented:
@robert-skarzycki I agree, current support for InnerCollection queries is far from being ideal. I'll try to revisit that after I finish my work on NodaTime support. But the fix won't be quick, probably no sooner than in version 3.6...
For now as a workaround I would suggest to use getters with flattened strucures. It's workaround as this will make data denormalized but it would work.

mysticmind commented:
@oskardudycz I have added `CollectionToArrayJsonConverter<T>` under `Marten.Util` and updated the unit tests to apply attribute on collection properties `[JsonConverter(typeof(Marten.Util.CollectionToArrayJsonConverter<TypeWithInnerCollections>))]`. All tests are green.

With this fix, we can close #1105 and #1192. We will also need to add some documentation on this.

oskardudycz commented:
@mysticmind I made few additional improvements:
- Refactored CollectionToArrayJsonConverter to not need generic parameter,
- added JsonNetContractResolver to allow global setting of storing collections as arrays

Could you have a look and share your thoughts? 😄

mysticmind commented:
@oskardudycz Looks like the tests are failing with some recent changes, can you please have a look?

oskardudycz commented:
@mysticmind should be good now. Too long document class name ;)

mysticmind commented on src/Marten/Util/CollectionToArrayJsonConverter.cs:
```diff
@@ -0,0 +1,44 @@
+    /// Note that without using custom `JsonConverter`, `Newtonsoft.Json` stores it as $type and $value.
+    /// Or you may need to resort to `Newtonsoft.Json.TypeNameHandling.None` which has its own side-effects
+    /// </summary>
+    /// <typeparam name="T">Type of the collection</typeparam>
```
Remove this document comment, not applicable any more

mysticmind commented on src/Marten/Util/CollectionToArrayJsonConverter.cs:
```diff
@@ -0,0 +1,44 @@
+    /// Or you may need to resort to `Newtonsoft.Json.TypeNameHandling.None` which has its own side-effects
+    /// </summary>
+    /// <typeparam name="T">Type of the collection</typeparam>
+    public class CollectionToArrayJsonConverter : JsonConverter
```
I think it would make sense to rename this class to `JsonNetCollectionToArrayJsonConverter` and move it under `Marten.Services.Json` namespace

mysticmind commented on src/Marten/Services/JsonNetSerializer.cs:
```diff
@@ -119,22 +123,27 @@ public Casing Casing
+        }
+
+        /// <summary>
+        /// Specify whether properties in the JSON document should use Camel or Pascal casing.
```
Looks like a copy & paste glitch, fix the xml comment ;-)

oskardudycz commented:
@mysticmind comments applied ;)
